### PR TITLE
Optimize tx index lookup in web3 rpc

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #760

## Description

Solution:
- emit tx index to cosmos events
- rpc side try to use the events, but fallback to heavier approach when fails.

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: rpc/ethereum/namespaces/eth/api.go [hunk 1]
```diff
@@ -664,75 +675,86 @@ func (e *PublicAPI) GetTransactionByHash(hash common.Hash) (*rpctypes.RPCTransac
 	return e.backend.GetTransactionByHash(hash)
 }
 
+// getTransactionByBlockHashAndIndex returns the transaction identified by hash and index.
```
Comment (yihuang) [inline rpc/ethereum/namespaces/eth/api.go:678]:
```suggestion
// getTransactionByBlockAndIndex is the common code shared by `GetTransactionByBlockNumberAndIndex` and `GetTransactionByBlockHashAndIndex`.
```

**Review by thomas-nguy — COMMENTED** [review]
lgtm

**Review by thomas-nguy — APPROVED** [review]

File: rpc/ethereum/backend/backend.go [hunk 2]
```diff
@@ -688,27 +689,57 @@ func (e *EVMBackend) GetTransactionByHash(txHash common.Hash) (*types.RPCTransac
...
+		for i := range msgs {
+			if msgs[i].Hash == hexTx {
+				txIndex = uint64(i)
+				found = true
+				break
+			}
+		}
+	}
+	if !found {
+		return nil, errors.New("Can't find index of ethereum tx")
```
Comment (fedekunze) [inline rpc/ethereum/backend/backend.go:737]:
```suggestion
		return nil, errors.New("can't find index of ethereum tx")
```

**Review by fedekunze — APPROVED** [review]
LGTM, can you also update the spec and add the new event?

**Comment by yihuang** [comment]
> LGTM, can you also update the spec and add the new event?

done

**Pull request merged**